### PR TITLE
Bluetooth: controller: Use entropy driver directly in bt_rand

### PR DESCRIPTION
**Pull request by wopu-ot**
Use entropy driver directly in bt_rand instead of stitching together
calls to sys_rand32_get to improve efficiency. The use of
sys_rand32_get could also leak timestamps into keys.

Signed-off-by: Wolfgang Puffitsch <wopu@demant.com>

### CODE REVIEW

**Review by thoh-ot — APPROVED** [review]

File: subsys/bluetooth/controller/crypto/crypto.c [hunk 1]
```diff
@@ -4,33 +4,33 @@
...
-			memcpy(buf8, &v, len);
-			break;
-		}
+	struct device *dev = entropy_driver;
+
+	if (unlikely(!dev)) {
+		/* Only one entropy device exists, so this is safe even
+		 * if the whole operation isn't atomic.
+		 */
+		dev = device_get_binding(CONFIG_ENTROPY_NAME);
```
Comment (carlescufi) [inline subsys/bluetooth/controller/crypto/crypto.c]: Could we cache `dev` somewhere to speed up execution? Maybe make it static?

File: subsys/bluetooth/controller/crypto/crypto.c [hunk 2]
```diff
@@ -4,33 +4,33 @@
...
-		if (len >= sizeof(v)) {
-			memcpy(buf8, &v, sizeof(v));
-
-			buf8 += sizeof(v);
-			len -= sizeof(v);
-		} else {
-			memcpy(buf8, &v, len);
-			break;
-		}
+	struct device *dev = entropy_driver;
```
Comment (carlescufi) [inline subsys/bluetooth/controller/crypto/crypto.c]: do you need the local variable?

Comment (wopu-ot) [reply subsys/bluetooth/controller/crypto/crypto.c]: Technically no. I basically copy/pasted that part from rand32_entropy_device.c

**Comment by wopu-ot** [comment]
Close and reopen to rerun CI

**Comment by Laczen** [comment]
@wopu-ot, nice proposal, does this not need some changes in Kconfig to select the entropy system ?

**Comment by wopu-ot** [comment]
@Laczen The controller already accesses the entropy device directly in other places and accordingly already selects ENTROPY_GENERATOR in the Kconfig.

**Comment by Laczen** [comment]
> @Laczen The controller already accesses the entropy device directly in other places and accordingly already selects ENTROPY_GENERATOR in the Kconfig.

OK

**Comment by cvinayak** [comment]
On nRF51, if bt_rand is used in the ISR, I would like to profile the CPU usage. Hence, give me time until 26/09/2019, after which feel free to merge on approvals from others.

**Comment by wopu-ot** [comment]
@carlescufi @cvinayak Do you have any further comments on this PR or can you approve?

**Review by carlescufi — APPROVED** [review]

**Review by aescolar — APPROVED** [review]

**Pull request merged**